File: hammer-border/src/main/java/dev/vini2003/hammer/border/impl/common/border/CubicWorldBorderMovingArea.java
package dev.vini2003.hammer.border.impl.common.border;

import net.minecraft.util.Util;
import net.minecraft.util.function.BooleanBiFunction;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.shape.VoxelShape;
import net.minecraft.util.shape.VoxelShapes;
import net.minecraft.world.border.WorldBorder;
import net.minecraft.world.border.WorldBorderStage;

public class CubicWorldBorderMovingArea implements CubicWorldBorderArea {
	private WorldBorder border;
	private CubicWorldBorder cubicBorder;
	
	private double oldSize;
	private double newSize;
	
	private double timeDuration;
	
	private long timeStart = 0L;
	private long timeEnd = 0L;
	
	public CubicWorldBorderMovingArea(WorldBorder border, double oldSize, double newSize, double timeDuration) {
		this.border = border;
		
		this.oldSize = oldSize;
		this.newSize = newSize;
		
		this.timeDuration = timeDuration;
	}
	
	@Override
	public double getBoundWest() {
		return MathHelper.clamp(border.getCenterX() - getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getBoundNorth() {
		return MathHelper.clamp(border.getCenterZ() - getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getBoundEast() {
		return MathHelper.clamp(border.getCenterX() + getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getBoundSouth() {
		return MathHelper.clamp(border.getCenterZ() + getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getBoundDown() {
		return MathHelper.clamp(cubicBorder.getCenterY() - border.getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getBoundUp() {
		return MathHelper.clamp(cubicBorder.getCenterY() + border.getSize() / 2.0D, (-border.getMaxRadius()), border.getMaxRadius());
	}
	
	@Override
	public double getSize() {
		var delta = (Util.getMeasuringTimeMs() - timeStart) / timeDuration;
		
		return delta < 1.0 ? MathHelper.lerp(delta, oldSize, newSize) : newSize;
	}
	
	@Override
	public double getShrinkingSpeed() {
		return Math.abs(oldSize - newSize) / (double) (timeEnd - timeStart);
	}
	
	@Override
	public long getSizeLerpTime() {
		return timeEnd - Util.getMeasuringTimeMs();
	}
	
	@Override
	public double getSizeLerpTarget() {
		return newSize;
	}
	
	@Override
	public WorldBorderStage getStage() {
		return newSize < oldSize ? WorldBorderStage.SHRINKING : WorldBorderStage.GROWING;
	}
	
	@Override
	public void onCenterChanged() {
	
	}
	
	@Override
	public void onMaxRadiusChanged() {
	
	}
	
	@Override
	public WorldBorder getBorder() {
		return border;
	}
	
	@Override
	public CubicWorldBorderArea getAreaInstance() {
		return getSizeLerpTime() <= 0L ? new CubicWorldBorderStaticArea(border, newSize) : this;
	}
	
	@Override
	public VoxelShape asVoxelShape() {
		return VoxelShapes.combineAndSimplify(
				VoxelShapes.UNBOUNDED,
				VoxelShapes.cuboid(
						Math.floor(getBoundWest()),
						Math.floor(getBoundDown()),
						Math.floor(getBoundNorth()),
						Math.ceil(getBoundEast()),
						Math.ceil(getBoundUp()),
						Math.ceil(getBoundSouth())
				),
				BooleanBiFunction.ONLY_FIRST
		);
	}
}

File: hammer-border/src/main/kotlin/dev/vini2003/hammer/border/impl/common/border/CubicWorldBorderMovingArea.kt
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.border.impl.common.border

import dev.vini2003.hammer.border.api.util.extension.centerY
import net.minecraft.util.Util
import net.minecraft.util.function.BooleanBiFunction
import net.minecraft.util.math.MathHelper
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.border.WorldBorder
import net.minecraft.world.border.WorldBorderStage
import kotlin.math.abs
import kotlin.math.ceil

class CubicWorldBorderMovingArea(override val border: WorldBorder, private val oldSize: Double, private val newSize: Double, private val timeDuration: Double) : CubicWorldBorderArea {
	private var timeEnd = 0L
	private var timeStart = 0L
	
	init {
		timeStart = Util.getMeasuringTimeMs()
		timeEnd = (timeStart + timeDuration).toLong()
	}
	
	override val boundWest: Double
		get() = MathHelper.clamp(
            border.centerX - size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val boundNorth: Double
		get() = MathHelper.clamp(
            border.centerZ - size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val boundEast: Double
		get() = MathHelper.clamp(
            border.centerX + size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val boundSouth: Double
		get() = MathHelper.clamp(
            border.centerZ + size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val boundDown: Double
		get() = MathHelper.clamp(
			border.centerY - border.size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val boundUp: Double
		get() = MathHelper.clamp(
            border.centerY + border.size / 2.0,
            (-border.maxRadius).toDouble(),
            border.maxRadius.toDouble()
        )
	
	override val size: Double
		get() {
			val delta = (Util.getMeasuringTimeMs() - timeStart).toDouble() / timeDuration
			
			if (delta < 1.0) {
				return MathHelper.lerp(delta, oldSize, newSize)
			} else {
				return newSize
			}
		}
	
	override val shrinkingSpeed: Double
		get() = abs(oldSize - newSize) / (timeEnd - timeStart).toDouble()
	
	override val sizeLerpTime: Long
		get() = timeEnd - Util.getMeasuringTimeMs()
	
	override val sizeLerpTarget: Double
		get() = newSize
	
	override val stage: WorldBorderStage
		get() {
			if (newSize < oldSize) {
				return WorldBorderStage.SHRINKING
			} else {
				return WorldBorderStage.GROWING
			}
		}
	
	override fun onCenterChanged() = Unit
	
	override fun onMaxRadiusChanged() = Unit
	
	override val areaInstance: CubicWorldBorderArea
		get() {
			if (sizeLerpTime <= 0L) {
				return CubicWorldBorderStaticArea(border, newSize)
			} else {
				return this
			}
		}
	
	override val voxelShape: VoxelShape
		get() = VoxelShapes.combineAndSimplify(
            VoxelShapes.UNBOUNDED,
            VoxelShapes.cuboid(
                ceil(boundWest),
                ceil(boundDown),
                ceil(boundNorth),
                ceil(boundEast),
                ceil(boundUp),
                ceil(boundSouth)
            ),
            BooleanBiFunction.ONLY_FIRST
        )
}